lib/route/cls/mall.c

```diff
@@ -0,0 +1,288 @@
+};
+
+/**
+ * @name Attribute Modifications
+ * @{
+ */
+
+int rtnl_mall_set_classid(struct rtnl_cls *cls, uint32_t classid)
+{
+	struct rtnl_mall *mall;
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	mall->m_classid = classid;
+	mall->m_mask |= MALL_ATTR_CLASSID;
+
+	return 0;
+}
+
+int rtnl_mall_get_classid(struct rtnl_cls *cls, uint32_t *classid)
+{
+	struct rtnl_mall *mall;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (!(mall->m_mask & MALL_ATTR_CLASSID))
+		return -NLE_INVAL;
+
+	*classid = mall->m_classid;
+	return 0;
+}
+
+int rtnl_mall_set_flags(struct rtnl_cls *cls, uint32_t flags)
+{
+	struct rtnl_mall *mall;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	mall->m_flags |= flags;
+	mall->m_mask |= MALL_ATTR_FLAGS;
+
+	return 0;
+}
+
+int rtnl_mall_get_flags(struct rtnl_cls *cls, uint32_t *flags)
+{
+	struct rtnl_mall *mall;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (!(mall->m_mask & MALL_ATTR_FLAGS))
+		return -NLE_INVAL;
+
+	*flags = mall->m_flags;
+	return 0;
+}
+
+int rtnl_mall_add_action(struct rtnl_cls *cls, struct rtnl_act *act)
+{
+	struct rtnl_mall *mall;
+
+	if (!act)
+		return 0;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	mall->m_mask |= MALL_ATTR_ACTION;
+	rtnl_act_get(act);
+	return rtnl_act_append(&mall->m_act, act);
+}
+
+struct rtnl_act *rtnl_mall_get_action(struct rtnl_cls *cls)
+{
+	struct rtnl_mall *mall;
+	struct rtnl_act *act;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return NULL;
+
+	if (!(mall->m_mask & MALL_ATTR_ACTION))
+		return NULL;
+
+	act = mall->m_act;
+	rtnl_act_get(act);
+
+	return act;
+}
+
+int rtnl_mall_del_action(struct rtnl_cls *cls, struct rtnl_act *act)
+{
+	struct rtnl_mall *mall;
+	int ret;
+
+	if (!act)
+		return 0;
+
+	if (!(mall = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (!(mall->m_mask & MALL_ATTR_ACTION))
+		return -NLE_INVAL;
+
+	ret = rtnl_act_remove(&mall->m_act, act);
+	if (ret)
+		return ret;
+
+	if (!mall->m_act)
+		mall->m_mask &= ~MALL_ATTR_ACTION;
+	rtnl_act_put(act);
+
+	return 0;
+}
+
+/** @} */
+
+static void mall_free_data(struct rtnl_tc *tc, void *data)
+{
+	struct rtnl_mall *mall = data;
+
+	if (mall->m_act)
+		rtnl_act_put_all(&mall->m_act);
+}
+
+static int mall_msg_parser(struct rtnl_tc *tc, void *data)
+{
+	struct rtnl_mall *mall = data;
+	struct nlattr *tb[TCA_MATCHALL_MAX + 1];
+	int err;
+
+	err = tca_parse(tb, TCA_MATCHALL_MAX, tc, mall_policy);
+	if (err < 0)
+		return err;
+
+	if (tb[TCA_MATCHALL_CLASSID]) {
+		mall->m_classid = nla_get_u32(tb[TCA_MATCHALL_CLASSID]);
+		mall->m_mask |= MALL_ATTR_CLASSID;
+	}
+
+	if (tb[TCA_MATCHALL_FLAGS]) {
+		mall->m_flags = nla_get_u32(tb[TCA_MATCHALL_FLAGS]);
+		mall->m_mask |= MALL_ATTR_FLAGS;
+	}
+
+	if (tb[TCA_MATCHALL_ACT]) {
+		mall->m_mask |= MALL_ATTR_ACTION;
+		err = rtnl_act_parse(&mall->m_act, tb[TCA_MATCHALL_ACT]);
```

Comment:
here, if the netlink message contains a `tb[TCA_MATCHALL_ACT]` attribute, but zero actions, then we will set `mall->m_mask |= MALL_ATTR_ACTION;`. Compare this to `rtnl_mall_del_action`, which clears `MALL_ATTR_ACTION` when removing the last action. It should be consistent: either, having no action means that the attribute flag is cleared, or once we set the attribute, we never clear it (meaning: that also an explicit empty list of actions, counts like actions are set). Probably, the second notion makes more sense, because in `mall_msg_fill` it determines whether to write a `TCA_MATCHALL_ACT` attribute.

Reply:
ok